kernel/src/main/scala/me/rexim/morganey/syntax/LambdaParser.scala

```diff
@@ -27,38 +28,65 @@ class LambdaParser extends JavaTokenParsers with ImplicitConversions {
   def variable: Parser[LambdaVar] =
     identifier.r ^^ { LambdaVar }
 
-  def numericLiteral: Parser[LambdaTerm] =
+  def validNumberLiteral: Parser[Int] =
     numberLiteral.r ^? ({
-      case IntMatcher(x) => ChurchNumberConverter.encodeNumber(x)
+      case IntMatcher(x) => x
     }, { (rawInt) => s"`$rawInt' is too big"})
 
-  def characterLiteral: Parser[LambdaTerm] = (
+  def numericLiteral: Parser[LambdaTerm] =
+    validNumberLiteral ^^ ChurchNumberConverter.encodeNumber
+
+  def validCharacterLiteral: Parser[Char] = (
     escapedCharLiteral.r ^^ { s =>
-      ChurchNumberConverter.encodeNumber(escapeSequences(s charAt 2))
-    }
-    | symbolCharLiteral.r ^^ { s =>
-      ChurchNumberConverter.encodeNumber(s charAt 1)
+      escapeSequences(s charAt 2)
     }
+    | symbolCharLiteral.r ^^ { _ charAt 1 }
   )
 
+  def characterLiteral: Parser[LambdaTerm] =
+    validCharacterLiteral ^^ { c => ChurchNumberConverter.encodeNumber(c.toInt) }
+
   def stringLiteralTerm: Parser[LambdaTerm] =
     stringLiteral ^^ { s =>
       ChurchPairConverter.encodeString(unquoteString(s))
     }
 
+  private def range[T](p: => Parser[T])(implicit ev: Integral[T]): Parser[LambdaTerm] = {
+    val q = p
+    val parser = q ~ opt(comma ~> q) ~ (rangeOperator ~> q)
+    parser ^^ { case start ~ next ~ exit =>
+      val step  = next.map(ev.minus(_, start)).getOrElse(ev.one)
+      val range = NumericRange.inclusive(start, exit, step).toList
+      val nums  = range map ev.toInt map ChurchNumberConverter.encodeNumber
+      ChurchPairConverter.encodeList(nums)
+    }
+  }
```

Comment:
> It's a optimization, which the parser can perform to a Morganey-program. This optimization could only be applied, if the bounds of the range are literals. If we keep this approach, it has to be tweaked to for example allow mixed usage of numbers and characters.

Sounds pretty good! I was thinking about introducing something like term coercing that looks like:

```scala
// Those are similar to stuff from Church Converters
def encode[A](a: A): LambdaTerm = ???
def decode[B](t: LambdaTerm): Option[B] = ???

def termCoerce[A, B](a: A): Option[B] = decode[B](encode[A](a))
```

so we can claim that the boundaries can only be the Scala types that are Term Coercible to number. I belive such notion can reduce the combinatorial complexity of such parsing optimization. :)

> You don't have to merge this, I can remove the ranges from the parser and we implement it later in a more Morganey-way.

I just thought, maybe it's not that critical... The code that uses these ranges is probably gonna be backward compatible with the Morganey ranges that I described above... Probably, I just overreacted. Let me just think about it for a while and I'm gonna answer here later today. Thanks for contributing to this language! :)